ipa-core/src/helpers/buffers/circular.rs

```diff
@@ -0,0 +1,788 @@
+    ///
+    /// No writes will be accepted after buffer is closed.
+    ///
+    /// ## Panics
+    /// if this buffer is already closed.
+    pub fn close(&mut self) {
+        debug_assert!(!self.closed, "Already closed");
+        self.closed = true;
+    }
+
+    /// Returns a handle that allows to perform a single write to the buffer. Write must be exactly
+    /// `write_size` bytes long and buffer must be open for writes and have sufficient capacity
+    /// to fit it. [`can_write`] can be used to check all of these conditions.
+    ///
+    /// ## Panics
+    /// If buffer is closed for writes or does not have enough capacity.
+    ///
+    /// [`can_write`]: Self::can_write
+    pub fn next(&mut self) -> Next<'_> {
+        debug_assert!(!self.closed, "Writing to a closed buffer");
+        debug_assert!(self.can_write(),
+            "Not enough space for the next write: only {av} bytes available, but at least {req} is required",
+            av = self.free(),
+            req = self.write_size
+        );
+
+        Next {
+            range: self.range(self.write, self.write_size),
+            buf: self,
+        }
+    }
+
+    /// Performs a read off this buffer. if [`can_read`] is false before reading,
+    /// this method will not panic, but will return an empty vector instead.
+    ///
+    /// if [`can_read`] is true before reading, this method is guaranteed to return
+    /// a non-empty vector. The length of it depends on whether this buffer is
+    /// closed or no. For closed buffers, the valid len will be in `[1, read_size]`
+    /// range, but always aligned with `write_size`. For open buffers, len
+    /// is always equal to `read_size`.
+    ///
+    /// [`can_read`]: Self::can_read
+    pub fn take(&mut self) -> Vec<u8> {
+        if !self.can_read() {
+            return Vec::new();
+        }
+
+        // Capacity is always a multiple of write_size, so delta is always aligned.
+        let delta = std::cmp::min(self.read_size, self.len());
+
+        let mut ret = Vec::with_capacity(delta);
+        let range = self.range(self.read, delta);
+
+        // If the read range wraps around, we need to split it
+        if range.end() < range.start() {
+            ret.extend_from_slice(&self.data[*range.start()..]);
+            ret.extend_from_slice(&self.data[..=*range.end()]);
+        } else {
+            ret.extend_from_slice(&self.data[range]);
+        }
+
+        self.read = self.inc(self.read, delta);
+
+        ret
+    }
+
+    /// Returns the number of bytes in this buffer.
+    pub fn len(&self) -> usize {
+        // Modulo arithmetic and wrapping/overflow rules in Rust
+        // make it difficult to write `(self.write - self.read) % 2*N`.
+        // It works well for power-of-two sizes, but for arbitrary
+        // buffer capacity, it is easier to use N - (a - b) because
+        // write is always ahead of read.
+        if self.write >= self.read {
+            self.wrap(self.write - self.read)
```

Comment:
the difference between write pointer and read pointer can never be more than `C`, but both can grow outside of [0, C] range because we clamp them at 2*C. The reason for mod 2*C is to be able to distinguish between full and empty buffer.